iup/all_objects.py

```diff
@@ -0,0 +1,366 @@
+import polars as pl
+import datetime as dt
+from sklearn.linear_model import LinearRegression
+import numpy as np
+import abc
+from typing_extensions import Self
+
+
+class UptakeData(pl.DataFrame, metaclass=abc.ABCMeta):
+    @abc.abstractmethod
+    def validate(self) -> None:
+        pass
+
+
+class IncidentUptakeData(UptakeData):
+    def validate(self):
+        assert set(self.columns).issuperset(
+            {"region", "date", "season", "elapsed", "interval", "estimate", "previous"}
```

Comment:
Here are some counterarguments that I think are worth discussing.
1) I think it's fine for IncidentUptakeData to have a very detailed structure, because it is never directly supplied by the user. The user supplies CumulativeUptakeData via `get_nis`. All these columns must be inferable from the CumulativeUptakeData, but not explicitly present there.
2) My "date" is equivalent to your "end of period," and my "interval" is the number of days since your "start of period." So ("date", "interval") and ("start of period", "end of period") contain the exact same information. And the former is better prepared to transform incident uptake into per-day incident uptake.
3) I only mean "region" in the generic sense of geography - it could be a state, a city, a county, etc. This is flexible to handle the goofy things that NIS does, like splitting Philadelphia County away from the Rest of PA.
4) A single uptake curve will certainly fit into IncidentUptakeData! In this case, the columns of "season" and "region" will each contain only one unique value repeated in every row (e.g. "2023" and "USA"), but that's still a sensible column. It may even prove useful, e.g. for generating figure plots and legends. I like having IncidentUptakeData always set up to handle maximum complexity (multiple regions, multiple seasons), even if it's not necessary in every instance. Carrying around a few single-value columns seems like small price to pay for never having to worry about extra contingencies when more complex data do come around.

Reply:
We talked in person; I don't think it's crazy to make an UptakeNIS class, and ensure that that has all the relevant columns. Although in that case, we might *also* want age, etc., whatever subpopulations are in the NIS data, which isn't great, because then you end up with having to do all this data validation that isn't really necessary.

I'm re-convincing myself that a suite of different data objects is the right one, but I won't be obstructive about that. Let's just have *something* validated, and we can figure out exact interfaces as we make more models.

Reply:
Sounds good! I am leaving "region" in the code for now, because it is helping me learn how to deal with grouping factors when manipulating the polars data frames. But we will keep in mind the possibility of more/different grouping factors, data objects with no grouping factors, etc. if the models call for them.